Fleshing out the document, added some words (#2)
autocrypt/protected-headers: draft-protected-headers.md

    email: dkg@fifthhorseman.net
informative:
normative:
 RFC2119:
 RFC2822:
 RFC3156:
 RFC5751:
 RFC8174:

>  RFC8551:

Reviewer: Add pEp's draft

```suggestion
 RFC8551:
 I-D.draft-luck-lamps-pep-header-protection-03:
```

Reviewer: We should also add draft-melnikov-smime-header-signing-05

--- abstract

> This document describes a common strategy to extend the cryptographic protections provided by PGP/MIME etc. to also protect message headers. In particular, how to encrypt the Subject line.

Reviewer: s/strategy/scheme/, since it's the word used in the rest of the document. 8551 and 3156 use "structure", 
Is it PGP/MIME used in other standards?, and cryptographic?. Both 8551 and 3156 use "secure" or "security".
Should we use e-mail or Email or message or Internet Message?

```suggestion
This document defines an schema to extend the cryptographic security provided by OpenPGP/MIME {{RFC3156}} to protect message headers in compatible ways with S/MIME {{RFC8551}} and {{I-D.draft-luck-lamps-pep-header-protection-03}}.
```
Is the compatibility part of the sentence right?.

Author: I like these changes - except I don't think what we are doing is actually compatible with S/MIME. Saying so implies we are compatible with whatever the current S/MIME spec has to say about header protection, which I am pretty sure is not true.

Protected Headers might be applicable to S/MIME as well, which is what I was trying to communicate, but since there are no actual implementations out there it's largely academic. To avoid confusion on that point, we might want to just omit it from the spec entirely, aside from maybe a dedicated "Future Work" section?

Reviewer: "Future work" sounds good. I actually need to read the other RFCs to have any opinion on the compatibility ;)

--- middle

Introduction
============

> The standards for cryptographic e-mail, PGP/MIME and S/MIME ({{RFC3156}} and {{RFC8551}}), can provide both integrity and confidentiality to the body of a MIME e-mail message. However, PGP/MIME alone provides no protection to message headers and the mechanism specified for header protection within S/MIME has not seen widespread adoption.

Reviewer: Repeat the references?
Signatures also provide authentication and non-repudiation, should we mention them?
Should we define what we mean by "protect/ed"?, and what 8551 and 3156 means by "secure/ity"?

```suggestion
MIME Security with OpenPGP and S/MIME standards can provide integrity, authentication, non-repudiation and confidentiality to the body of a MIME e-mail message. However, OpenPGP/MIME alone does not protect message headers and the structure to protect headers defined in S/MIME has not seen widespread adoption.
```

> This document defines a scheme, "Protected Headers for Cryptographic E-mail", which has been adopted by multiple existing e-mail clients in order to extend the cryptographic protections provided by PGP/MIME to also protect the {{RFC2822}} message headers.

Reviewer: Maybe mention the clients at the end of the sentence?.
Is it usual in RFC to mention concrete implementations?

```suggestion
This document defines a scheme, "Protected Headers for Cryptographic E-mail" to extend the cryptographic protections provided by OpenPGP/MIME to also protect the message headers {{RFC2822}}, which have been already adopted by multiple e-mail clients.
```

Author: I don't know if it's usual to mention implementations, but it's definitely a very important point for the overall "standardizability" of the draft. Maybe @dkg can comment on whether this is the right place for that?

Reviewer 2: I think we generally should avoid mentioning specific implementations in the draft, unless it's in the Acknowledgements section.  In discussion around the draft: yes, definitely we'll want to mention specific implementations.

> In particular, we describe how to encrypt the Subject line and how to preserve backwards compatibility so an encrypted subject remains available to recipients using software that does not implement support for the Protected Headers scheme.

Reviewer: s/line/header/?

```suggestion
In particular, we describe how to encrypt the Subject line and how to preserve backwards compatibility so that an encrypted subject remains available to recipients using software that does not implement support for the Protected Headers scheme.
```

We also discuss some of the compatibility constraints and usability concerns which motivated the design of the scheme, as well as limitations.

> The authors believe the technique is broadly applicable would also apply to other MIME-compatible cryptographic e-mail systems, including S/MIME.  Furthermore, this technique has already proven itself as a useful building block for other improvements to cryptographic e-mail, such as the Autocrypt Level 1.1 "Gossip" mechanism.

Reviewer: How to add reference to Autocrypt?

```suggestion
The authors believe that this scheme would also apply to other MIME-compatible cryptographic e-mail systems, including S/MIME.  Furthermore, this scheme is already a building block for other improvements to cryptographic e-mail, such as the Autocrypt Level 1.1 "Gossip" mechanism.
```

Requirements Language
---------------------

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in BCP 14 {{RFC2119}} {{RFC8174}} when, and only when, they appear in all capitals, as shown here.

Terminology
-----------

> 

Reviewer: Looking at memoryhole draft, it seems that the format of this type of enumeration list is something like:
I also think that definitions do not include "is"/"are".

```
MUA
: stands for Mail User Agent, an e-mail client.
```

For the purposes of this document, we define the following concepts:

   * *MUA* is short for Mail User Agent; an e-mail client.

>    * *Protection* of message data refers to cryptographic encryption and/or signatures, providing authenticity, confidentiality or both.

Reviewer: Mention here again other signatures properties?

>    * *Cryptographic Envelope* is all MIME structure directly dictated by the cryptographic e-mail system in use.

Reviewer: I find this sentence a bit confusing. Is the "cryptographic e-mail system" S/MIME and OpenPGP/MIME?.
I don't like the word "dictated" :P

Author: I'm trying to be abstract here, and I agree it makes things confusing. The cryptographic e-mail system could be OpenPGP/MIME, or S/MIME, or some other imaginary cryptosystem for MIME e-mail. See my other comment; I think maybe we should just stop trying to be abstract and make this concretely about OpenPGP/MIME, and add a future work section that proposes it might also be useful in other contexts.

I inherited the abstract direction from my interpretation of @dkg 's intent, so I'd like his feedback on whether we change course. It would simplify things though.

Reviewer 2: I'm fine with the first draft focusing on PGP/MIME, and i think that will help us keep the document more narrowly targeted :)

It is important to me that we don't document it as though it could not be compatible with S/MIME though, and leave the door open to a future draft including S/MIME as well.

   * *Cryptographic Payload* is all message data protected by the Cryptographic Envelope.
   * *Original Headers* are the RFC2822 message headers as known to the sending MUA at the time of message composition.
   * *Protected Headers* are any headers protected by the scheme described in this document.
   * *Exposed Headers* are any headers outside the Cryptographic Payload (protected or not).

>    * *Obscured Headers* are any headers which have been modified or removed from the set of Exposed Headers.

Reviewer: What about "Modified Headers" instead of "Obscured"?

Author: I prefer obscured, because that communicates the intent of WHY we are modifying them. We're not modifying them just-because, we are doing so to reduce exposure of sensitive data. Happy to hear other points of view though.

Reviewer: Agree with your point

>    * *Legacy Display Part* is a MIME construct which guarantees visibility of data from the Original Headers which may have been removed or obscured from the Unprotected Headers.

Reviewer: s/MIME construct/MIME structure/?

Author: It's not just structure, it's also content. Thus "construct." But I agree it's a bit weird.

> The Cryptographic Envelope fully encloses the Cryptographic Payload, whether the message is signed or encrypted. The Original Headers, aside from Content-Type headers which directly pertain to the cryptographic structure, are considered to be outside of both.

Reviewer: Not sure if mentioning here the Original Headers adds more confusion.
Do you mean that Content-Type is contained in the Cryptographic Envelope? Maybe the diagrams should start to be added here, or start explaining the different cases.

```suggestion
The Cryptographic Envelope contains the Cryptographic Payload, both in the case of signed and encrypted messages. The Original Headers are the headers outside of both the Cryptographic Envelope and Payload.
```

Author: Using the word "contains" may to some readers imply that the payload is part of the envelope. I am pretty sure @dkg wanted to keep the concepts as separate as possible, which is why I wrote "fully enclose".

In the second half, the goal was not to redefine the Original Headers (they were already defined), but explain how they relate to the other concepts. So I disagree with that change too. :-)

Reviewer: ok, i think i've not understood what we call Envelope, maybe pointing it the full examples will help.

Reviewer 2: I have some drafted sections that are much more explicit about what i think we should call the Cryptographic Payload and Cryptographic Envelope.

I'd be happy to try to apply those sections on top of this work.

Aside from that limitation, this specification does not at this time define or limit the methods a MUA may use to convert Exposed Headers into Obscured Headers.

Legacy Display
--------------

Message Interpretation
======================

(Brief discussion about potential strategies to reverse the process above.)

Examples
========

(Use diagrams from Juga to explain the envelope/payload concept)

(Add sample messages using this syntax, where E or S tell us which bits are part of the Cryptographic Payload.)

    . From: Alice Lovelace <alice@example.org>
    . To: Bob Babbage <bob@example.org>
    . Subject: ...
    . Content-Type: multipart/encrypted; protocol=...; boundary=1234
    .
    . --1234
    . Content-Type: application/pgp-encrypted
    .
    . Version: 1
    .
    . --1234
    . Content-Type: application/octet-stream; name="encrypted.asc"
    .
    E Content-Type: multipart/mixed; boundary=ABCD
    E (copied headers)
    E
    E --ABCD
    E (Legacy display part)
    E
    E --ABCD
    E (original message)
    E --ABCD--
    . --1234--

> 

Reviewer: Hmm, what if we use complete examples to make it clearer and follow the same structures as in the diagrams?.
Here a proposal:
```suggestion
Text in `[` and `]` is not part of the message or not as it is, it's a clarification in the examples.
Message-ID and boundary strings have been replaced by `message_id` and `multipart_X_boundary` to make the examples more clear.
Extra new lines have been added to also make the examples more clear.

TODO: Note what we call Cryptographic Envelope in the examples.
FIXME: Enigmail adds `; protected-headers="v1"` to `multipart/mixed`. Is that needed?

Example OpenPGP signed message
-------------------------------

In which the body is "text/plain".

    From: Alice Lovelace <alice@openpgp.example>
    To: Bob Babbage <bob@openpgp.example>
    Subject: Hi
    Message-ID: <message_id>
    Date: Thu, 17 Oct 2019 06:09:00 +0000
    MIME-Version: 1.0
    Content-Type: multipart/signed; micalg=pgp-sha512;
     protocol="application/pgp-signature";
     boundary="multipart_signed_boundary"

    This is an OpenPGP/MIME signed message (RFC 4880 and 3156)

    --multipart_signed_boundary

    [This is the payload containing the headers]
    Content-Type: multipart/mixed; boundary="multipart_mixed_boundary"
    From: Alice Lovelace <alice@openpgp.example>
    To: Alice Lovelace <alice@openpgp.example>
    Message-ID: <message_id>
    Subject: Hi

    --multipart_mixed_boundary

    Content-Type: text/plain; charset=utf-8
    Content-Language: en-US
    Content-Transfer-Encoding: quoted-printable

    Hi Bob!,

    we have a meeting today.

    Best,
    Alice.

    --multipart_mixed_boundary--

    --multipart_signed_boundary
    Content-Type: application/pgp-signature; name="signature.asc"
    Content-Description: OpenPGP digital signature
    Content-Disposition: attachment; filename="signature.asc"

    -----BEGIN PGP SIGNATURE-----

    [Signature of the multipart/mixed structure.]

    -----END PGP SIGNATURE-----

    --multipart_signed_boundary--

Example OpenPGP encrypted message
----------------------------------

In which the body of the message is "text/plain".

    From: Alice Lovelace <alice@openpgp.example>
    To: Bob Babbage <bob@openpgp.example>
    Message-ID: <message_id>
    Date: Thu, 17 Oct 2019 06:09:00 +0000
    MIME-Version: 1.0
    Subject: ...
    Content-Type: multipart/encrypted;
     protocol="application/pgp-encrypted";
     boundary="multipart_encrypted_boundary"

    This is an OpenPGP/MIME encrypted message (RFC 4880 and 3156)
    --multipart_encrypted_boundary
    Content-Type: application/pgp-encrypted
    Content-Description: PGP/MIME version identification

    Version: 1

    --multipart_encrypted_boundary
    Content-Type: application/octet-stream; name="encrypted.asc"
    Content-Description: OpenPGP encrypted message
    Content-Disposition: inline; filename="encrypted.asc"

    -----BEGIN PGP MESSAGE-----

    [Encrypted payload]
    -----END PGP MESSAGE-----

    --multipart_encrypted_boundary--

After decrypting the Cryptographic Payload, this is the payload containing the 
headers:

TODO: would be the decrypted part just "text/plain" with the headers on top of 
the body?

Example encrypted message in the "compat" form
-----------------------------------------------

In which the body of the message is "text/plain".
Note that the outer structure of the message is the same as in the example above.

    From: Alice Lovelace <alice@openpgp.example>
    To: Bob Babbage <bob@openpgp.example>
    Message-ID: <message_id>
    Date: Thu, 17 Oct 2019 06:09:00 +0000
    MIME-Version: 1.0
    Subject: ...
    Content-Type: multipart/encrypted;
     protocol="application/pgp-encrypted";
     boundary="multipart_encrypted_boundary"

    This is an OpenPGP/MIME encrypted message (RFC 4880 and 3156)
    --multipart_encrypted_boundary
    Content-Type: application/pgp-encrypted
    Content-Description: PGP/MIME version identification

    Version: 1

    --multipart_encrypted_boundary
    Content-Type: application/octet-stream; name="encrypted.asc"
    Content-Description: OpenPGP encrypted message
    Content-Disposition: inline; filename="encrypted.asc"

    -----BEGIN PGP MESSAGE-----

    [Encrypted payload]
    -----END PGP MESSAGE-----

    --multipart_encrypted_boundary--

After decrypting the Cryptographic Payload, this is the payload containing the 
headers:

    Content-Type: multipart/mixed; boundary="multipart_mixed_boundary"
    From: Alice Lovelace <alice@openpgp.example>
    To: Bob Babbage <bob@openpgp.example>
    Message-ID: <message_id>
    Subject: Hi

    --multipart_mixed_boundary
    Content-Type: text/rfc822-headers
    Content-Disposition: inline

    [This is the legacy display]
    From: Alice Lovelace <alice@openpgp.example>
    To: Alice Lovelace <alice@openpgp.example>
    Subject: subject, signed and encrypted
    [FIXME: Include Message-ID also here (enigmail does not)?]

    --multipart_mixed_boundary
    Content-Type: text/plain; charset=utf-8

    Hi Bob!,

    we have a meeting today.

    Best,
    Alice.
    --multipart_mixed_boundary--

Example multilayer cryptographic envelope
-----------------------------------------

In which the body of the message is "text/plain".
Note that the outer structure of the message is the same as in the example above.

    From: Alice Lovelace <alice@openpgp.example>
    To: Bob Babbage <bob@openpgp.example>
    Message-ID: <message_id>
    Date: Thu, 17 Oct 2019 06:09:00 +0000
    MIME-Version: 1.0
    Subject: ...
    Content-Type: multipart/encrypted;
     protocol="application/pgp-encrypted";
     boundary="multipart_encrypted_boundary"

    This is an OpenPGP/MIME encrypted message (RFC 4880 and 3156)
    --multipart_encrypted_boundary
    Content-Type: application/pgp-encrypted
    Content-Description: PGP/MIME version identification

    Version: 1

    --multipart_encrypted_boundary
    Content-Type: application/octet-stream; name="encrypted.asc"
    Content-Description: OpenPGP encrypted message
    Content-Disposition: inline; filename="encrypted.asc"

    -----BEGIN PGP MESSAGE-----

    [Encrypted payload]
    -----END PGP MESSAGE-----

    --multipart_encrypted_boundary--

After decrypting the Cryptographic Payload, this is the payload containing the 
headers:
Note that this structure is the same as in the example signed message.

    From: Alice Lovelace <alice@openpgp.example>
    To: Bob Babbage <bob@openpgp.example>
    Subject: Hi
    Message-ID: <message_id>
    Date: Thu, 17 Oct 2019 06:09:00 +0000
    MIME-Version: 1.0
    Content-Type: multipart/signed; micalg=pgp-sha512;
     protocol="application/pgp-signature";
     boundary="multipart_signed_boundary"

    This is an OpenPGP/MIME signed message (RFC 4880 and 3156)

    --multipart_signed_boundary

    [This is the payload containing the headers]
    Content-Type: multipart/mixed; boundary="multipart_mixed_boundary"
    From: Alice Lovelace <alice@openpgp.example>
    To: Alice Lovelace <alice@openpgp.example>
    Message-ID: <message_id>
    Subject: Hi

    --multipart_mixed_boundary

    Content-Type: text/plain; charset=utf-8
    Content-Language: en-US
    Content-Transfer-Encoding: quoted-printable

    Hi Bob!,

    we have a meeting today.

    Best,
    Alice.

    --multipart_mixed_boundary--

    --multipart_signed_boundary
    Content-Type: application/pgp-signature; name="signature.asc"
    Content-Description: OpenPGP digital signature
    Content-Disposition: attachment; filename="signature.asc"

    -----BEGIN PGP SIGNATURE-----

    [Signature of the multipart/mixed structure.]

    -----END PGP SIGNATURE-----

    --multipart_signed_boundary--

```

Author: I'm going to commit this, I think it's a valuable chunk of text to iterate on. Thanks for doing this!

I still prefer a unified display with the column to one side to explain which parts are encrypted and which are not - I dislike having to scroll up and down to compare different versions of the same message. I also feel we should omit as much as possible to help people more easily get an overview of what actually matters. For example the content-transfer-encoding usually doesn't matter at all. Including it is therefore just noise, in my opinion.

But different brains work differently. If you feel this is clearer, I believe that it will indeed be clearer for some readers. I am not sure how best to balance that!

Author: (To clarify: those headers all matter in real life. But many of them do not matter within the context of how Protected Headers work, so I feel omitting them is a good way to reduce noise in the spec.)
